reactor-core/src/main/java/reactor/core/publisher/FluxWindowTimeout.java

```diff
@@ -26,52 +26,1590 @@
+		}
+
+		@Override
+		public void subscribe(CoreSubscriber<? super T> actual) {
+			long previousState = markSubscribedOnce(this);
+			if (hasSubscribedOnce(previousState)) {
+				Operators.error(actual, new IllegalStateException("Only one subscriber allowed"));
+				return;
+			}
+
+			this.actual = actual;
+
+			actual.onSubscribe(this);
+
+			previousState = markSubscriberSet(this);
+			if (isFinalized(previousState) || hasWorkInProgress(previousState)) {
+				return;
+			}
+
+			if (!hasValues(previousState) && isTerminated(previousState)) {
+				final Throwable t = this.error;
+				if (t != null) {
+					actual.onError(t);
+				} else {
+					actual.onComplete();
+				}
+			}
+		}
+
+		@Override
+		public CoreSubscriber<? super T> actual() {
+			return this.actual;
+		}
+
+		@Override
+		public void request(long n) {
+			Operators.addCap(REQUESTED, this, n);
+			final long previousState = markHasRequest(this);
+
+			if (hasWorkInProgress(previousState) || isCancelled(previousState) || isFinalized(previousState)) {
+				return;
+			}
+
+			if (hasValues(previousState)) {
+				drain((previousState | HAS_SUBSCRIBER_SET_STATE) + 1);
+			}
+		}
+
+		@Override
+		public void cancel() {
+			final long previousState = markCancelled(this);
+			if (isCancelled(previousState) || isFinalized(previousState) || hasWorkInProgress(previousState)) {
+				return;
+			}
+
+			clearAndFinalize();
+		}
+
+		/**
+		 * This method is called from the parent when the parent is cancelled. We don't
+		 * send any exceptions to the subscriber here but we deliver complete window
+		 * and discard all further values delivered through the {@link #sendNext} method
+		 *
+		 * @return previous state
+		 */
+		long sendCancel() {
+			for (;;) {
+				final long state = this.state;
+
+				if (isCancelledByParent(state)) {
+					return state;
+				}
+
+				final long cleanState = state & ~WORK_IN_PROGRESS_MAX;
+				final long nextState = cleanState |
+						TERMINATED_STATE |
+						PARENT_CANCELLED_STATE |
+						(
+							hasSubscriberSet(state)
+								? hasValues(state)
+									? incrementWork(state & WORK_IN_PROGRESS_MAX)
+									: FINALIZED_STATE
+								: 0
+						);
+
+				if (STATE.compareAndSet(this, state, nextState)) {
+
+					if (isFinalized(state)) {
+						return state;
+					}
+
+					if (!isTimeout(state)) {
+						final Disposable timer = TIMER.getAndSet(this, DISPOSED);
+						if (timer != null) {
+							timer.dispose();
+						}
+					}
+
+					if (hasSubscriberSet(state)) {
+						if (hasWorkInProgress(state)) {
+							return state;
+						}
+
+						if (isCancelled(state)) {
+							clearAndFinalize();
+							return state;
+						}
+
+						if (hasValues(state)) {
+							drain(nextState);
+						} else {
+							this.actual.onComplete();
+						}
+					}
+
+					return state;
+				}
+			}
+		}
+
+		boolean sendNext(T t) {
+			int received = this.received + 1 ;
+			this.received = received;
+
+			this.queue.offer(t);
```

Comment:
should we guard against an already cancelled window here and directly discard each element before they get a chance to be added to the queue?